File: PokedexObjC/PokedexObjC/ViewControllers/PokemonDetailViewController.m
//
//  PokemonDetailViewController.m
//  PokedexObjC
//
//  Created by Carolyn Lea on 10/12/18.
//  Copyright © 2018 Carolyn Lea. All rights reserved.
//

#import "PokemonDetailViewController.h"

@interface PokemonDetailViewController ()

@end

void *KVOContext = &KVOContext;

@implementation PokemonDetailViewController

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    self.nameLabel.text = self.pokemon.name;
    self.title = self.pokemon.name;
    
    [PokemonController.sharedController fillInDetailsFor:self.pokemon];
}


-(void)updateViews
{
    dispatch_async(dispatch_get_main_queue(), ^{
        
        self.idLabel.text = [NSString stringWithFormat: @"%ld", (long)self.pokemon.identifier];
        self.abilitiesLabel.text = [NSString stringWithFormat: @"%@", self.pokemon.abilities];
        NSData * imageData = [[NSData alloc] initWithContentsOfURL: self.pokemon.spriteURL];
        
        self.imageView.image = [UIImage imageWithData: imageData];
        
    });
}

- (void)setPokemon:(Pokemon *)pokemon
{
    if (pokemon != _pokemon)
    {
        
        [_pokemon removeObserver:self forKeyPath:@"sprite" context:KVOContext];
        [_pokemon removeObserver:self forKeyPath:@"identifier" context:KVOContext];
        [_pokemon removeObserver:self forKeyPath:@"abilities" context:KVOContext];
        _pokemon = pokemon;
        
        [_pokemon addObserver:self forKeyPath:@"sprite" options:NSKeyValueObservingOptionInitial context:KVOContext];
        [_pokemon addObserver:self forKeyPath:@"identifier" options:NSKeyValueObservingOptionInitial context:KVOContext];
        [_pokemon addObserver:self forKeyPath:@"abilities" options:NSKeyValueObservingOptionInitial context:KVOContext];
    }
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
    if (context != KVOContext)
    {
        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
        return;
    }
    
    
    [self updateViews];
    
}


@end

File: PokedexObjC/PokedexObjC/ViewControllers/PokemonDetailViewController.swift
//
//  PokemonDetailViewController.swift
//  PokedexObjC
//
//  Created by Benjamin Hakes on 3/8/19.
//  Copyright © 2019 Benjamin Hakes. All rights reserved.
//

import UIKit

class PokemonDetailViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        guard let objectToObserve = objectToObserve else {fatalError(" No object to observe")}
        observation = objectToObserve.observe(\.speciesName) {object,change in
            print("Object Change Observed:\(object):\(change)")
            self.updateViews()
        }
        
        guard let pokemonAPI = pokemonAPI else { fatalError("No pokemonAPI exists.")}
        
        pokemonAPI.fillInDetails(for: objectToObserve)
        
        self.updateViews()
        
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        observation?.invalidate()
    }

    // MARK: Private Methods
    
    func updateViews(){
        
        guard let pokemon = objectToObserve else { return }
        DispatchQueue.main.async {
            self.nameLabel.text = pokemon.name
            self.speciesLabel.text = pokemon.speciesName
            self.idLabel.text = pokemon.identifier
            
            guard let images = pokemon.images else { return }
            
            var imageIndex = self.segmentedControl.selectedSegmentIndex
            if imageIndex == 0 {
                imageIndex = 1
            } else {
                imageIndex = 0
            }
            let imageURL = images[imageIndex]
            do {
                let data = try Data(contentsOf: imageURL)
                self.imageView.image = UIImage(data: data)
            } catch {
                print("could not update image view")
            }
            
            guard let abilities = pokemon.abilities else { return }
            
            self.abilitiesTextView.text = ""
            for ability in abilities {
                
                self.abilitiesTextView.text += "\(ability.name)\n"
            }
            
        }
        reloadInputViews()
        
    }
    
    @IBAction func didTapSegmentedControl(_ sender: UISegmentedControl) {
        updateViews()
        
    }
    

    // MARK: Properties
    
    @IBOutlet weak var segmentedControl: UISegmentedControl!
    @IBOutlet weak var nameLabel: UILabel!
    @IBOutlet weak var speciesLabel: UILabel!
    @IBOutlet weak var idLabel: UILabel!
    @IBOutlet weak var abilitiesTextView: UITextView!
    @IBOutlet weak var imageView: UIImageView!
   
    @objc var pokemonAPI: PokemonAPI?
    @objc dynamic var observation: NSKeyValueObservation?
    @objc dynamic var objectToObserve: BHPokemon?

}
